zebra: rib_process_dplane_results hook call

dlqs opened the pull request:
To be done: command to set script name

polychaeta reviewed (changes requested):
Thanks for your contribution to FRR!

* One of your commits has a missing or badly formatted `Signed-off-by` line; we can't accept your contribution until all of your commits have one
* One of your commits does not have a blank line between the summary and body; this will break `git log --oneline`

If you are a new contributor to FRR, please see our [contributing guidelines](http://docs.frrouting.org/projects/dev-guide/en/latest/workflow.html#coding-practices-style).

After making changes, you do not need to create a new PR. You should perform an [amend or interactive rebase](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History) followed by a [force push](https://git-scm.com/docs/git-push#Documentation/git-push.txt---force).

dlqs commented on lib/frrlua.c:
```diff
@@ -295,6 +311,57 @@ void *lua_tointegerp(lua_State *L, int idx)
+	lua_setfield(L, -2, "backup_num");
+	lua_pushinteger(L, *(nexthop->backup_idx));
+	lua_setfield(L, -2, "backup_idx");
+	if (nexthop->nh_encap_type == NET_VXLAN) {
```
check later

dlqs commented on lib/frrlua.c:
```diff
@@ -295,6 +311,57 @@ void *lua_tointegerp(lua_State *L, int idx)
+	lua_setfield(L, -2, "weight");
+	lua_pushinteger(L, nexthop->backup_num);
+	lua_setfield(L, -2, "backup_num");
+	lua_pushinteger(L, *(nexthop->backup_idx));
```
check later

dlqs commented on zebra/zebra_script.c:
```diff
@@ -0,0 +1,364 @@
+		break;
+	case DPLANE_OP_SYS_ROUTE_ADD:
+	case DPLANE_OP_SYS_ROUTE_DELETE:
+		// nothing to encode
```
check later

dlqs commented on lib/frrlua.c:
```diff
@@ -223,6 +223,22 @@ void *lua_toin6addr(lua_State *L, int idx)
+void lua_pushethaddr(lua_State *L, const struct ethaddr *addr)
+{
+	lua_newtable(L);
+	lua_pushinteger(L, *(addr->octet));
```
check later

dlqs commented on lib/frrlua.c:
```diff
@@ -295,6 +311,57 @@ void *lua_tointegerp(lua_State *L, int idx)
+	lua_setfield(L, -2, "type");
+	lua_pushinteger(L, nexthop->flags);
+	lua_setfield(L, -2, "flags");
+	if (nexthop->type == NEXTHOP_TYPE_BLACKHOLE) {
```
check later

polychaeta reviewed (changes requested):
Thanks for your contribution to FRR!

* One of your commits has a missing or badly formatted `Signed-off-by` line; we can't accept your contribution until all of your commits have one
* One of your commits does not have a blank line between the summary and body; this will break `git log --oneline`

If you are a new contributor to FRR, please see our [contributing guidelines](http://docs.frrouting.org/projects/dev-guide/en/latest/workflow.html#coding-practices-style).

After making changes, you do not need to create a new PR. You should perform an [amend or interactive rebase](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History) followed by a [force push](https://git-scm.com/docs/git-push#Documentation/git-push.txt---force).

dlqs commented on lib/frrlua.c:
```diff
@@ -295,6 +311,57 @@ void *lua_tointegerp(lua_State *L, int idx)
+	lua_newtable(L);
+	struct nexthop *nexthop;
+	int i = 0;
+	for (ALL_NEXTHOPS_PTR(ng, nexthop)) {
+		lua_pushnexthop(L, nexthop);
+		lua_seti(L, -2, i);
+		i++;
+	}
```
check later

this encodes the nexthops with integer keys

polychaeta reviewed (dismissed):
Thanks for your contribution to FRR!

* One of your commits has a missing or badly formatted `Signed-off-by` line; we can't accept your contribution until all of your commits have one
* One of your commits does not have a blank line between the summary and body; this will break `git log --oneline`

If you are a new contributor to FRR, please see our [contributing guidelines](http://docs.frrouting.org/projects/dev-guide/en/latest/workflow.html#coding-practices-style).

After making changes, you do not need to create a new PR. You should perform an [amend or interactive rebase](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History) followed by a [force push](https://git-scm.com/docs/git-push#Documentation/git-push.txt---force).

polychaeta reviewed (changes requested):
Thanks for your contribution to FRR!

* One of your commits has a missing or badly formatted `Signed-off-by` line; we can't accept your contribution until all of your commits have one
* One of your commits does not have a blank line between the summary and body; this will break `git log --oneline`

If you are a new contributor to FRR, please see our [contributing guidelines](http://docs.frrouting.org/projects/dev-guide/en/latest/workflow.html#coding-practices-style).

After making changes, you do not need to create a new PR. You should perform an [amend or interactive rebase](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History) followed by a [force push](https://git-scm.com/docs/git-push#Documentation/git-push.txt---force).

polychaeta reviewed (dismissed):
Thanks for your contribution to FRR!

* One of your commits has a missing or badly formatted `Signed-off-by` line; we can't accept your contribution until all of your commits have one
* One of your commits does not have a blank line between the summary and body; this will break `git log --oneline`

If you are a new contributor to FRR, please see our [contributing guidelines](http://docs.frrouting.org/projects/dev-guide/en/latest/workflow.html#coding-practices-style).

After making changes, you do not need to create a new PR. You should perform an [amend or interactive rebase](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History) followed by a [force push](https://git-scm.com/docs/git-push#Documentation/git-push.txt---force).

polychaeta reviewed (changes requested):
Thanks for your contribution to FRR!

* One of your commits has an improperly formatted commit message
* One of your commits has a missing or badly formatted `Signed-off-by` line; we can't accept your contribution until all of your commits have one
* One of your commits does not have a blank line between the summary and body; this will break `git log --oneline`

If you are a new contributor to FRR, please see our [contributing guidelines](http://docs.frrouting.org/projects/dev-guide/en/latest/workflow.html#coding-practices-style).

After making changes, you do not need to create a new PR. You should perform an [amend or interactive rebase](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History) followed by a [force push](https://git-scm.com/docs/git-push#Documentation/git-push.txt---force).

polychaeta reviewed (changes requested):
Thanks for your contribution to FRR!

* One of your commits has an improperly formatted commit message
* One of your commits has a missing or badly formatted `Signed-off-by` line; we can't accept your contribution until all of your commits have one
* One of your commits does not have a blank line between the summary and body; this will break `git log --oneline`

If you are a new contributor to FRR, please see our [contributing guidelines](http://docs.frrouting.org/projects/dev-guide/en/latest/workflow.html#coding-practices-style).

After making changes, you do not need to create a new PR. You should perform an [amend or interactive rebase](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History) followed by a [force push](https://git-scm.com/docs/git-push#Documentation/git-push.txt---force).

polychaeta reviewed (changes requested):
Thanks for your contribution to FRR!

* One of your commits has a missing or badly formatted `Signed-off-by` line; we can't accept your contribution until all of your commits have one
* One of your commits does not have a blank line between the summary and body; this will break `git log --oneline`

If you are a new contributor to FRR, please see our [contributing guidelines](http://docs.frrouting.org/projects/dev-guide/en/latest/workflow.html#coding-practices-style).

After making changes, you do not need to create a new PR. You should perform an [amend or interactive rebase](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History) followed by a [force push](https://git-scm.com/docs/git-push#Documentation/git-push.txt---force).

sworleys commented on zebra/zebra_router.h:
```diff
@@ -36,6 +37,36 @@ extern "C" {
+DECLARE_HASH(script_names, struct script_names_entry, item,
+	     zebra_script_hash_cmp, zebra_script_hash_key);
+
+#endif /* HAVE_SCRIPTING */
```
So... presumably other daemons are gonna need some facility like this other than just zebra. Is it possible for us to abstract most of this into lib/ so a daemon can more easily add this functionality itself?

Look at `lib/nexthop_group.c`  and it's `vty` commands are an example of this. Basically a daemon can register callbacks to handle cli commands from the user. This way you create the cli in lib, and each daemon can maintain it's own hash table of the function names it's interested in.

sworleys replied:
Marking resolved since we went the route of keeping individual hookpoint calls but moving all the hashing/table into lib/.

polychaeta reviewed (dismissed):
Thanks for your contribution to FRR!

* One of your commits has a missing or badly formatted `Signed-off-by` line; we can't accept your contribution until all of your commits have one
* One of your commits does not have a blank line between the summary and body; this will break `git log --oneline`

If you are a new contributor to FRR, please see our [contributing guidelines](http://docs.frrouting.org/projects/dev-guide/en/latest/workflow.html#coding-practices-style).

After making changes, you do not need to create a new PR. You should perform an [amend or interactive rebase](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History) followed by a [force push](https://git-scm.com/docs/git-push#Documentation/git-push.txt---force).

sworleys commented on zebra/zebra_vty.c:
```diff
@@ -4298,6 +4299,26 @@ DEFUN_HIDDEN(no_zebra_kernel_netlink_batch_tx_buf,
+      "Specify a script file for a hook call\n"
+      "Hook call function name\n"
+      "Script name (same as filename in /etc/frr/scripts/, without .lua)\n")
+{
```
For the command, I think @qlyoung  wanted to leave it up to individual places to add the function/script info rather than an all-encompassing script command per daemon, right?

so like:

```
zebra on-dplane-install script myfunnyscript.lua
```

qlyoung replied:
Correct, we shouldn't need this command defined; the idea is to accept script names naturally where we would other configuration related to the function the script would serve, not to have a global mapping of script -> hook point.

dlqs replied:
updated hook call specific command `zebra on-rib-process script SCRIPT`

polychaeta reviewed (dismissed):
Thanks for your contribution to FRR!

* One of your commits has a missing or badly formatted `Signed-off-by` line; we can't accept your contribution until all of your commits have one
* One of your commits does not have a blank line between the summary and body; this will break `git log --oneline`

<details>
<summary><b>Click for style suggestions</b></summary>

To apply these suggestions:

<p>

```
curl -s https://gist.githubusercontent.com/polychaeta/48cee0aa67dbbcf12af670406cac5f1b/raw/797b31752843e9e2e536a240d98b094239377171/cr_9440_1630346122.diff | git apply
```

</p>

<p>

```diff
diff --git a/zebra/zebra_router.c b/zebra/zebra_router.c
index 0e92a1d73..28fd105a7 100644
--- a/zebra/zebra_router.c
+++ b/zebra/zebra_router.c
@@ -32,7 +32,6 @@
 #include "debug.h"
 
 
-
 DEFINE_MTYPE_STATIC(ZEBRA, RIB_TABLE_INFO, "RIB table info");
 DEFINE_MTYPE_STATIC(ZEBRA, ZEBRA_RT_TABLE, "Zebra VRF table");
 

```

</p>
</details>


If you are a new contributor to FRR, please see our [contributing guidelines](http://docs.frrouting.org/projects/dev-guide/en/latest/workflow.html#coding-practices-style).

After making changes, you do not need to create a new PR. You should perform an [amend or interactive rebase](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History) followed by a [force push](https://git-scm.com/docs/git-push#Documentation/git-push.txt---force).

qlyoung commented on doc/user/scripting.rst:
```diff
@@ -26,3 +36,50 @@ In order to use scripting, FRR must be built with ``--enable-scripting``.
+
+Scripts live in :file:`/etc/frr/scripts/` by default. This is
+configurable at compile time via ``--with-scriptdir``. It may be
+overriden at runtime with the ``--scriptdir`` daemon option.
```
IIRC you use vim - ^v, select the paragraph, `gqq` will reflow a paragraph for an 80 column line limit. I get 
```rst
Scripts live in :file:`/etc/frr/scripts/` by default. This is configurable at
compile time via ``--with-scriptdir``. It may be overriden at runtime with the
``--scriptdir`` daemon option.
```
Not a blocking comment though, just a nit.

dlqs replied:
Ah. I reflowed the file

qlyoung commented on doc/user/scripting.rst:
```diff
@@ -26,3 +36,50 @@ In order to use scripting, FRR must be built with ``--enable-scripting``.
+function was created in `/etc/frr/scripts/my_dplane_script.lua`,
+the following vtysh command sets the script for the hook call:
+
+.. code-block:: shell
```
Not a shell script, leave off the syntax specifier

dlqs replied:
updated in https://github.com/FRRouting/frr/pull/9440/commits/7c4d8a01a50ee0da66295815ce82d5b31357978f

dlqs replied:
seems like rst demands a arg for `code-block` so I changed to console

qlyoung commented on doc/user/scripting.rst:
```diff
@@ -26,3 +36,50 @@ In order to use scripting, FRR must be built with ``--enable-scripting``.
+
+The documentation for :ref:`on-rib-process-dplane-results` indicates that the
+``script`` command should be used to set the script. Assuming that the above
+function was created in `/etc/frr/scripts/my_dplane_script.lua`,
```
```
:file:`/etc/frr/scripts/my_dplane_script.lua`
```

dlqs replied:
updated in https://github.com/FRRouting/frr/pull/9440/commits/7c4d8a01a50ee0da66295815ce82d5b31357978f

qlyoung commented on doc/user/scripting.rst:
```diff
@@ -26,3 +36,50 @@ In order to use scripting, FRR must be built with ``--enable-scripting``.
+
+After the script is set, when the hook call is hit, FRR will look for a
+*on_rib_process_dplane_results* function in
+`/etc/frr/scripts/my_dplane_script.lua` and run it with the ``ctx`` object
```
```
:file:`/etc/frr/scripts/my_dplane_script.lua`
```

dlqs replied:
updated in https://github.com/FRRouting/frr/pull/9440/commits/7c4d8a01a50ee0da66295815ce82d5b31357978f

qlyoung commented on doc/user/zebra.rst:
```diff
@@ -1334,3 +1334,203 @@ Debugging
+const struct zebra_dplane_ctx
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+.. code-block:: lua
```
Not lua, drop the specifier

dlqs replied:
updated in https://github.com/FRRouting/frr/pull/9440/commits/7c4d8a01a50ee0da66295815ce82d5b31357978f

qlyoung commented on doc/user/zebra.rst:
```diff
@@ -1334,3 +1334,203 @@ Debugging
+const struct nh_grp
+^^^^^^^^^^^^^^^^^^^
+
+.. code-block:: lua
```
Not lua, drop the specifier

dlqs replied:
updated in https://github.com/FRRouting/frr/pull/9440/commits/7c4d8a01a50ee0da66295815ce82d5b31357978f

qlyoung commented on lib/frrscript.c:
```diff
@@ -32,6 +32,84 @@
+int frrscript_names_add_function_name(const char *function_name)
+{
+	struct frrscript_names_entry *insert =
+		XCALLOC(MTYPE_TMP, sizeof(*insert));
```
This should be a new `MTYPE`, not `MTYPE_TMP`

dlqs replied:
updated to `MTYPE_SCRIPT` which is already an MTYPE for frrscript - let me know if I should make a whole new one instead

qlyoung commented on lib/frrscript.h:
```diff
@@ -24,13 +24,51 @@
+extern void lua_pushzebra_dplane_ctx(lua_State *L,
+				     const struct zebra_dplane_ctx *ctx);
+extern void lua_decode_zebra_dplane_ctx(lua_State *L, int idx,
+					struct zebra_dplane_ctx *ctx);
```
These should go in the `extern "C"` block

dlqs replied:
updated in https://github.com/FRRouting/frr/pull/9440/commits/730d6378ff9ec8cfca23f117c3d91310f14be3df

qlyoung commented on zebra/zebra_router.c:
```diff
@@ -31,6 +31,8 @@
 #include "debug.h"
 
+
+
```
Remove this

dlqs replied:
updated in https://github.com/FRRouting/frr/pull/9440/commits/a2df35f1459b554daff2c41507d16c66d1c0de50

qlyoung commented on zebra/zebra_router.h:
```diff
@@ -46,6 +47,7 @@ struct zebra_router_table {
 
 	struct route_table *table;
 };
+
```
Remove this

dlqs replied:
updated in https://github.com/FRRouting/frr/pull/9440/commits/a5819304e72c8fee7521fab8aa64852b75871c7b

qlyoung commented on zebra/zebra_script.c:
```diff
@@ -0,0 +1,398 @@
+ * You should have received a copy of the GNU General Public License
+ * along with FRR; see the file COPYING.  If not, write to the Free
+ * Software Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
+ * 02111-1307, USA.
```
This is the wrong copyright header
http://docs.frrouting.org/projects/dev-guide/en/latest/workflow.html#source-file-header

dlqs replied:
updated in 575a205

qlyoung commented on zebra/zebra_script.h:
```diff
@@ -0,0 +1,45 @@
+ * You should have received a copy of the GNU General Public License
+ * along with FRR; see the file COPYING.  If not, write to the Free
+ * Software Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
+ * 02111-1307, USA.
```
Wrong header
http://docs.frrouting.org/projects/dev-guide/en/latest/workflow.html#source-file-header

dlqs replied:
updated in 575a205

qlyoung commented:
Please fix poly's errors as well

dlqs commented on doc/user/zebra.rst:
```diff
@@ -1334,3 +1334,203 @@ Debugging
+
+**Arguments**
+
+* :ref:`const struct zebra_dplane_ctx<const-struct-zebra-dplane-ctx>` ctx
```
Yup, shows as a link for me:

![image](https://user-images.githubusercontent.com/31354724/132085051-f1f15adb-ea8d-4375-994f-f9465e16dfe5.png)

polychaeta reviewed (changes requested):
Thanks for your contribution to FRR!

* One of your commits has a missing or badly formatted `Signed-off-by` line; we can't accept your contribution until all of your commits have one
* One of your commits does not have a blank line between the summary and body; this will break `git log --oneline`

<details>
<summary><b>Click for style suggestions</b></summary>

To apply these suggestions:

<p>

```
curl -s https://gist.githubusercontent.com/polychaeta/5600b7c856eee11140c06a6865d8cd1d/raw/797b31752843e9e2e536a240d98b094239377171/cr_9440_1630743237.diff | git apply
```

</p>

<p>

```diff
diff --git a/zebra/zebra_router.c b/zebra/zebra_router.c
index 0e92a1d73..28fd105a7 100644
--- a/zebra/zebra_router.c
+++ b/zebra/zebra_router.c
@@ -32,7 +32,6 @@
 #include "debug.h"
 
 
-
 DEFINE_MTYPE_STATIC(ZEBRA, RIB_TABLE_INFO, "RIB table info");
 DEFINE_MTYPE_STATIC(ZEBRA, ZEBRA_RT_TABLE, "Zebra VRF table");
 

```

</p>
</details>


If you are a new contributor to FRR, please see our [contributing guidelines](http://docs.frrouting.org/projects/dev-guide/en/latest/workflow.html#coding-practices-style).

After making changes, you do not need to create a new PR. You should perform an [amend or interactive rebase](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History) followed by a [force push](https://git-scm.com/docs/git-push#Documentation/git-push.txt---force).

dlqs commented on doc/user/zebra.rst:
```diff
@@ -1334,3 +1334,203 @@ Debugging
+     * integer ucast_probes
+     * integer mcast_probes
+
+   * table gre (if zd_op is DPLANE_OP_GRE_SET**
```
updated in https://github.com/FRRouting/frr/pull/9440/commits/7c4d8a01a50ee0da66295815ce82d5b31357978f

dlqs commented on doc/user/scripting.rst:
```diff
@@ -1,22 +1,32 @@
+.. seealso:: Developer docs for scripting
+
+How to use
+==========
```
yup, updated in e3d5ed9

polychaeta reviewed (changes requested):
Thanks for your contribution to FRR!

* One of your commits has a missing or badly formatted `Signed-off-by` line; we can't accept your contribution until all of your commits have one
* One of your commits does not have a blank line between the summary and body; this will break `git log --oneline`

If you are a new contributor to FRR, please see our [contributing guidelines](http://docs.frrouting.org/projects/dev-guide/en/latest/workflow.html#coding-practices-style).

After making changes, you do not need to create a new PR. You should perform an [amend or interactive rebase](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History) followed by a [force push](https://git-scm.com/docs/git-push#Documentation/git-push.txt---force).

polychaeta reviewed (changes requested):
Thanks for your contribution to FRR!

* One of your commits has a missing or badly formatted `Signed-off-by` line; we can't accept your contribution until all of your commits have one

If you are a new contributor to FRR, please see our [contributing guidelines](http://docs.frrouting.org/projects/dev-guide/en/latest/workflow.html#coding-practices-style).

After making changes, you do not need to create a new PR. You should perform an [amend or interactive rebase](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History) followed by a [force push](https://git-scm.com/docs/git-push#Documentation/git-push.txt---force).

dlqs commented:
@qlyoung @sworleys 
Made a few of significant changes (and it looks messy since I rebased a bit, sorry about that) since the last review, so I'll summarize them here:

 - changed top-level script command to hook-specific script command: `zebra on-rib-process script SCRIPT`
 - added frrscript_names_destroy to deallocate the names struct
 - text and formatting corrections

qlyoung commented:
@frrbot rereview

sworleys commented:
@dlqs I think I am happy with where this is. I think the docs could still be improved but Idk how to make them better at the moment :upside_down_face:  so hopefully when we get more hooks that will become clear. Maybe some way to collapse the arguments or something. Just UI stuff really, there is no need to iterate on them.

Can you rebase and force push? Once CI tests are happy I will merge it in.

sworleys commented:
> # Continuous Integration Result: SUCCESSFUL
> Congratulations, this patch passed basic tests
> 
> Tested-by: NetDEF / OpenSourceRouting.org CI System
> 
> CI System Testrun URL: https://ci1.netdef.org/browse/FRR-PULLREQ2-222/
> 
> This is a comment from an automated CI system.
> For questions and feedback in regards to this CI system, please feel free to email
> Martin Winter - mwinter (at) opensourcerouting.org.
> 
> ### Warnings Generated during build:
> Checkout code: Successful with additional warnings
> ```
.....


@dlqs  Could you address these before I push in?

dlqs commented:
@sworleys I've given it a cleanup!

dlqs commented:
Rebased it because of conflicts. Take a look @qlyoung @sworleys ?
